File: src/ReverseLinkedList.java
/**
 * Created by hermanwu on 2/22/18.
 */
public class ReverseLinkedList {
    public ListNode reverseList(ListNode head) {
        if (head == null) {
            return head;
        }

        ListNode newHead = null;

        while (head != null) {
            ListNode temp = head.next;
            head.next = newHead;
            newHead = head;
            head = temp;
        }

        return newHead;
    }
}

File: data-structure/LinkedList/reverse-linked-list/reverse-linked-list.ts
class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

export function reverseLinkedList(head: ListNode | null): ListNode | null {
  let tail = null;

  while (head) {
    const temp = head;
    head = head.next;
    temp.next = tail;
    tail = temp;
  }
  return tail;
}

const mockNode = new ListNode(1, new ListNode(2, new ListNode(3)));
const mockSingleNode = new ListNode(1);
const mockNull = null;
console.log(mockNode);

console.log(reverseLinkedList(mockNode));
console.log(reverseLinkedList(mockSingleNode));
console.log(reverseLinkedList(mockNull));
